server/src/main/java/org/opensearch/index/seqno/LocalCheckpointTracker.java

```diff
@@ -131,22 +131,49 @@ public synchronized void markSeqNoAsProcessed(final long seqNo) {
      * @param seqNo the sequence number to mark as persisted
      */
     public synchronized void markSeqNoAsPersisted(final long seqNo) {
-        markSeqNo(seqNo, persistedCheckpoint, persistedSeqNo);
+        markSeqNo(seqNo, persistedCheckpoint, persistedSeqNo, false);
     }
 
-    private void markSeqNo(final long seqNo, final AtomicLong checkPoint, final LongObjectHashMap<CountedBitSet> bitSetMap) {
+    /**
+     * Updates the processed checkpoint to the provided sequence number if segment replication is enabled
+     *
+     * @param seqNo the sequence number to mark as processed
+     */
+    public synchronized void fastForwardProcessedSeqNo(final long seqNo) {
+        markSeqNo(seqNo, processedCheckpoint, null, true);
+    }
+
+    private boolean shouldUpdateSeqNo(final long seqNo, boolean segrep, final AtomicLong checkpoint) {
+        return !((seqNo <= checkpoint.get()) || (segrep && seqNo > persistedCheckpoint.get()));
+    }
+
+    @SuppressForbidden(reason = "Object#notifyAll")
+    private void markSeqNo(
+        final long seqNo,
+        final AtomicLong checkPoint,
+        final LongObjectHashMap<CountedBitSet> bitSetMap,
+        boolean segrep
+    ) {
         assert Thread.holdsLock(this);
         // make sure we track highest seen sequence number
         advanceMaxSeqNo(seqNo);
-        if (seqNo <= checkPoint.get()) {
+        if (shouldUpdateSeqNo(seqNo, segrep, checkPoint) == false) {
             // this is possible during recovery where we might replay an operation that was also replicated
             return;
         }
-        final CountedBitSet bitSet = getBitSetForSeqNo(bitSetMap, seqNo);
-        final int offset = seqNoToBitSetOffset(seqNo);
-        bitSet.set(offset);
-        if (seqNo == checkPoint.get() + 1) {
-            updateCheckpoint(checkPoint, bitSetMap);
+        if (segrep) {
```

Comment:
I realize that this was done in response to my comment, but this may be too far in the direction of unification 😄 

What do you think of the following structure?

1. Instead of branching on this boolean value, have this method return the boolean from `shouldUpdateSeqNo` (or in-line that method's logic here). 
2. Have a single method that executes the `else` part of this - both `markSeqNoAsPersisted` and `markSeqNoAsProcessed` can invoke this method. This will cover the existing code flows.
3. Finally, move the `if` clause portion directly  into `fastForwardProcessedSeqNo` - this will cover the seg-rep code flow.

Reply:
shouldUpdateSeqNo doesn't have the logic to return a boolean representing segrep, but followed the other suggestions to have separate code flows